### ISSUE TITLE
Personalised QR code not accepted by app, invalid

### ISSUE THREAD

**Issue by maryisdead** [corona-warn-app/cwa-quicktest-onboarding#30]
## Describe the bug

QR code for personalised transmission to CWA is reported as being invalid.

## Expected behaviour

QR code should be accepted by the app.

## Steps to reproduce the issue

Sample implementation below. Two QR codes are being generated, personalised and pseudonymised.

The pseudonymised code works just fine, the personalised one will not be accepted. Link also does not work.

Is there something simple I'm missing? I quadruple-checked everything.

```php
<?php

/* composer.json
{
    "name": "acme/cwa",
    "require": {
        "chillerlan/php-qrcode": "^3.4",
        "guzzlehttp/guzzle": "^7.3"
    }
}
*/

error_reporting(-1);
ini_set('display_errors', 'On');

require __DIR__ . '/vendor/autoload.php';

use chillerlan\QRCode\QRCode;
use chillerlan\QRCode\QROptions;
use GuzzleHttp\Client;

function create_qr_code($data, $options = ['scale' => 10]) {
    $options = new QROptions($options);
    $qr_code = (new QRCode($options))->render($data);

    return $qr_code;
}

function guidv4() {
    $data = PHP_MAJOR_VERSION < 7 ? openssl_random_pseudo_bytes(16) : random_bytes(16);
    $data[6] = chr(ord($data[6]) & 0x0f | 0x40);    // Set version to 0100
    $data[8] = chr(ord($data[8]) & 0x3f | 0x80);    // Set bits 6-7 to 10
    
    return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
};

$test_scan_url_base = 'https://s.coronawarn.app?v=1#';
$secret = '1Au+y[?qhC)>h.fL++&QYd:blA{5qL;mrb-7-|I.NQGh::=zC]R`>tE0bDH$uQA_';

$timestamp = time();
$salt = strtoupper(bin2hex(random_bytes(16)));


// Personalized

$dob = '1981-07-25';
$first_name = 'Max';
$last_name = 'Mustermann';
$id = guidv4();

$hash_personalized = hash_hmac(
    'sha256',
    sprintf('%s#%s#%s#%s#%s#%s', $dob, $first_name, $last_name, $timestamp, $id, $salt),
    $secret
);

$payload_personalized = [
    'fn'        => $first_name,
    'ln'        => $last_name,
    'dob'       => $dob,
    'timestamp' => $timestamp,
    'testid'    => $id,
    'salt'      => $salt,
    'hash'      => $hash_personalized,
];

$json_payload_personalized = json_encode($payload_personalized, JSON_UNESCAPED_UNICODE);
// The following doesn't work either, just in case their docs are serious about the whitespace.
// (Though the pseudo variant works with regular JSON, so …)
/*
$json_payload = sprintf(
    '{ "fn": "%s", "ln": "%s", "dob": "%s", "timestamp": %d, "testid": "%s", "salt": "%s", "hash": "%s" }',
    $first_name,
    $last_name,
    $dob,
    $timestamp,
    $id,
    $salt,
    $hash
);
*/
$base64_payload_personalized = base64_encode($json_payload_personalized);
$test_scan_url_personalized = $test_scan_url_base . $base64_payload_personalized;
$test_scan_qr_code_personalized = create_qr_code($test_scan_url_personalized);


// Pseudo

$hash_pseudo = hash_hmac(
    'sha256',
    sprintf('%s#%s', $timestamp, $salt),
    $secret
);

$payload_pseudo = [
    'timestamp' => $timestamp,
    'salt'      => $salt,
    'hash'      => $hash_pseudo,
];

$json_payload_pseudo = json_encode($payload_pseudo, JSON_UNESCAPED_UNICODE);
$base64_payload_pseudo = base64_encode($json_payload_pseudo);
$test_scan_url_pseudo = $test_scan_url_base . $base64_payload_pseudo;
$test_scan_qr_code_pseudo = create_qr_code($test_scan_url_pseudo);

?>
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>CWA</title>
</head>
<body>

    <h1>Personalisiert</h1>

    <p>
        <a href="<?= $test_scan_url_personalized ?>" target="_blank">Link</a>
    </p>

    <p>
        <img src="<?= $test_scan_qr_code_personalized ?>" width=300 alt>
    </p>

    <h1>Pseudonymisiert</h1>

    <p>
        <a href="<?= $test_scan_url_pseudo ?>" target="_blank">Link</a>
    </p>

    <p>
        <img src="<?= $test_scan_qr_code_pseudo ?>" width=300 alt>
    </p>
    
</body>
</html>
```

**Comment by vaubaehn** [comment]
Hi, looks like you're encoding in base64, but base64url is required. Could you check again with correct encoding?

**Comment by maryisdead** [comment]
@vaubaehn Thank you for your input! Thought that wasn't the actual issue I added it in for safety.

The actual issue was that I generated the hash using a secret. I now simply run `hash('sha256', $data)` instead of `hash_hmac('sha256', $data, $secret)`.

I'm not sure why the HMAC worked for pseudonymised tests in the first place. Probably stricter checks on their end when there's personalised data.
